Add Interoperable protocol and implement tool import from multiple frameworks

davorrunje opened the pull request:
<!-- Thank you for your contribution! Please review https://ag2ai.github.io/ag2/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

This PR introduces several new functionality to the AG2:
1. New class `Tool` with methods `register_for_execution` and  `register_for_llm`.

2. The new `Interoperable` protocol to enable interoperability with other frameworks and  the new `Interoperability` class to expose interoperability classes implementing the `Interoperable` protocol. The following frameworks are supported for tool import:
   - [x] CrewAI
   - [x] LangChain
   - [x] Pydantic AI
    
### Install
Available extras are: `interop-crewai`, `interop-langchain` and `interop-pydantic`.

An example on how to install support forCrewAI tools:
```
pip install ag2[interop-crewai]
```

### Usage

An example on how to convert `FileReadTool` and register it for execution and LLM:

```python
from crewai_tools import 

# create CrewAI tool
crewai_tool = FileReadTool()

# converts CrewAI tool into AG2 Tool
tool = Interoperability.convert_tool(type="crewai", tool=crewai_tool)

# register AG2 tool for LLM and execution
tool.register_for_execution(user_proxy)
tool.register_for_llm(chatbot)
```

The list of supported frameworks can be dynamically retrieved using the `Interoperability.get_supported_types` method:

```python
supported_types = Interoperability.get_supported_types()
print(supported_types)
# expected output
# ["crewai", "langchain", "pydanticai"]
```

## Related issue number

<!-- For example: "Closes #1234" -->

Closes #208 

## Checks

- [x] I've included any doc changes needed for https://ag2ai.github.io/ag2/. See https://ag2ai.github.io/ag2/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

qingyun-wu commented:
@davorrunje thanks for the PR. It seems all checks are failing. Please fix. 

comparing to
```
crewai_tool = FileWriterTool()
crewai_interop = CrewAIInteroperability()
ag2_tool = crewai_interop.convert_tool(crewai_tool)

ag2_tool.register_for_execution(user_proxy)
ag2_tool.register_for_llm(chatbot)
```

is the following a better interface? 

```
crewai_tool = FileWriterTool()
interop = Interoperability()
ag2_tool = interop.convert_tool(tool=crewai_tool, type='crewai')

ag2_tool.register_for_execution(user_proxy)
ag2_tool.register_for_llm(chatbot)
```

marklysze commented on autogen/interoperability/crewai/__init__.py:
```diff
@@ -0,0 +1,17 @@
+import sys
+
+if sys.version_info < (3, 10) or sys.version_info >= (3, 13):
+    raise ImportError("This submodule is only supported for Python versions 3.10, 3.11, and 3.12")
```
Is there a reason we can't support 3.13? AG2 Python is up to 3.13, now.

davorrunje replied:
CrewAI does not support 3.13. However, I refactored that part so importing submodules will not raise an exception, but a class with missing imports or an unsupported Python version would not be available in the interface of the `Interoperability` class.

AgentGenie commented on autogen/tools/tool.py:
```diff
@@ -0,0 +1,47 @@
+from autogen.agentchat.conversable_agent import ConversableAgent
+
+try:
+    from crewai.tools import BaseTool as CrewAITool
```
This is not used. 
Supposedly it should not be specific to CrewAITool, right?

davorrunje replied:
Yes, this was a copy-paste error. Thanx!

AgentGenie commented on autogen/tools/tool.py:
```diff
@@ -0,0 +1,47 @@
+__all__ = ["Tool"]
+
+
+class Tool:
```
Doc string for the interface?

davorrunje replied:
Added.

AgentGenie commented on autogen/interoperability/interoperability.py:
```diff
@@ -0,0 +1,12 @@
+
+
+@runtime_checkable
+class Interoperable(Protocol):
```
Doc string for the interface?

davorrunje replied:
Added.

AgentGenie commented on autogen/coding/jupyter/embedded_ipython_code_executor.py:
```diff
@@ -13,6 +13,7 @@
 from queue import Empty
 from typing import Any, ClassVar, List
 
+import jupyter_kernel_gateway
```
Not used?

davorrunje replied:
removed

AgentGenie commented on autogen/interoperability/crewai/crewai.py:
```diff
@@ -0,0 +1,41 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+from typing import Any, cast
```
cast not used

davorrunje replied:
Removed 👍

AgentGenie commented on autogen/interoperability/interoperability.py:
```diff
@@ -0,0 +1,12 @@
+
+@runtime_checkable
+class Interoperable(Protocol):
+    def convert_tool(self, tool: Any) -> Tool: ...
```
This looks like a class method to me, isn't it?
And cls method would simplify the code like
`
crewai_tool = FileWriterTool()
crewai_interop = CrewAIInteroperability()
ag2_tool = CrewAIInteroperability.convert_tool(crewai_tool)
`
to 
`
ag2_tool = CrewAIInteroperability.convert_tool(FileWriterTool())
`

davorrunje replied:
You are right, it has been changed now. Thanx!

davorrunje commented:
> @davorrunje thanks for the PR. It seems all checks are failing. Please fix.

Fixed.

We added LangChain and are working on Pydantic AI support now.

> 
> comparing to
> 
> ```
> crewai_tool = FileWriterTool()
> crewai_interop = CrewAIInteroperability()
> ag2_tool = crewai_interop.convert_tool(crewai_tool)
> 
> ag2_tool.register_for_execution(user_proxy)
> ag2_tool.register_for_llm(chatbot)
> ```
> 
> is the following a better interface?
> 
> ```
> crewai_tool = FileWriterTool()
> interop = Interoperability()
> ag2_tool = interop.convert_tool(tool=crewai_tool, type='crewai')
> 
> ag2_tool.register_for_execution(user_proxy)
> ag2_tool.register_for_llm(chatbot)
> ```
> 
> And we should add other types in convert_tool later.
> 
> @sonichi please take a look too!

It's a question of aesthetics; different people will have different opinions. For example, frameworks such as Keras and PyTorch use the style you proposed extensively, and it works quite well. We can easily support it if you wish.

AgentGenie commented on autogen/interop/crewai/crewai.py:
```diff
@@ -0,0 +1,41 @@
+    return re.sub(r"\W|^(?=\d)", "_", s)
+
+
+class CrewAIInteroperability(Interoperable):
```
Interoperable is a Protocol, do we still need inheritance here?

AgentGenie commented on autogen/tools/pydantic_ai_tool.py:
```diff
@@ -0,0 +1,29 @@
+        }
+
+    def register_for_llm(self, agent: ConversableAgent) -> None:
+        agent.update_tool_signature(self._func_schema, is_remove=False)
```
I am curious why this tool is different?

rjambrecic replied:
CrewAI and LangChain tools have have `args_schema` parameter within the object:
- Pydantic model class to validate and parse the tool's input arguments. 

So we can create a function with the correct typing
```python
def func(tool_input: langchain_tool.args_schema) -> Any: 
       return langchain_tool.run(tool_input.model_dump())
```
And parameters schema will be taken from the function signature.

So we can register tool for LLM like:
```python
def register_for_llm(self, agent: ConversableAgent) -> None:
      agent.register_for_llm(name=self._name, description=self._description)(self._func)
```
__________________

PydanticAI tools don't have the `args_schema`, they have `_parameters_json_schema
So the easiest way to prepare the tool config for the LLM is following
```python
self._func_schema = {
    "type": "function",
    "function": {
        "name": tool.name,
        "description": tool.description,
        "parameters": tool._parameters_json_schema,
    },
}
```

In this case, we want to `self._func_schema` instead of parsing the function signature, so we need to override `register_for_llm`:
```python
def register_for_llm(self, agent: ConversableAgent) -> None:
      agent.update_tool_signature(self._func_schema, is_remove=False)
```

AgentGenie commented:
Nice PR!
One comment in general, could you please add doc string to interfaces and APIs?

davorrunje commented:
> Nice PR! One comment in general, could you please add doc string to interfaces and APIs?

Thanx :)

We added docs at the end once the API stabilized. We'll polish it a bit more tomorrow before the release.

marklysze reviewed (approved):
Great PR, thanks, let's go!
